indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java

```diff
@@ -360,6 +369,23 @@ private LockGranularityDetermineResult determineSegmentGranularity(List<DataSegm
     }
   }
 
+  static boolean isGuaranteedRollup(IndexIOConfig ioConfig, IndexTuningConfig tuningConfig)
+  {
+    Preconditions.checkState(
+        !tuningConfig.isForceGuaranteedRollup() || !ioConfig.isAppendToExisting(),
```

Comment:
or, pls add a comment explaining why, technically, we can't append new perfectly rolled up segments to existing. Is this something not possible or just not enabled at this point.

Reply:
Hmm, good point. I think, technically, we can append perfectly-rolled up set of segments to non-perfectly rolled up one even though some code should be fixed to support it (ex, `HashBasedNumberedShardSpec` always assumes that the start partitionId in the perfectly-rolled up set is 0). I'm not sure how useful it is though. Maybe it could be useful in some use cases. Added javadoc.

Reply:
Added some comment.